service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/ReactorRunner.java

```diff
@@ -0,0 +1,88 @@
+/*
+ *  Copyright (c) Microsoft. All rights reserved.
+ *  Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+package com.microsoft.azure.sdk.iot.service.transport.amqps;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.qpid.proton.Proton;
+import org.apache.qpid.proton.engine.BaseHandler;
+import org.apache.qpid.proton.engine.HandlerException;
+import org.apache.qpid.proton.reactor.Reactor;
+
+import java.io.IOException;
+
+@Slf4j
+public class ReactorRunner
+{
+    private final static String THREAD_NAME_PREFIX = "azure-iot-sdk-ReactorRunner-";
+    private final String threadName;
+    private final Reactor reactor;
+    public static final int REACTOR_TIMEOUT = 3141; // reactor timeout in milliseconds
+
+    public ReactorRunner(BaseHandler baseHandler, String threadNamePostfix) throws IOException
+    {
+        this.reactor = Proton.reactor(baseHandler);
+        this.threadName = THREAD_NAME_PREFIX + threadNamePostfix;
+    }
+
+    public void run(long timeoutMs)
+    {
+        Thread.currentThread().setName(this.threadName);
+
+        try
+        {
+            log.trace("Starting reactor thread {}", this.threadName);
+            this.reactor.setTimeout(REACTOR_TIMEOUT);
+
+            long startTime = System.currentTimeMillis();
+            long endTime = startTime + timeoutMs;
+
+            this.reactor.start();
+            while (this.reactor.process())
+            {
+                if (System.currentTimeMillis() > endTime) break;
+            }
+
+            this.reactor.stop();
+            this.reactor.process();
+        }
+        catch (HandlerException e)
+        {
+            log.debug("Encountered an exception while running reactor on thread {}", threadName, e);
+        }
+        finally
+        {
+            log.trace("Freeing reactor now that reactor thread is done");
+            this.reactor.free();
+        }
+
+        log.trace("Finished reactor thread {}", this.threadName);
+    }
+
+    public void run()
+    {
+        Thread.currentThread().setName(this.threadName);
+
+        try
+        {
+            log.trace("Starting reactor thread {}", this.threadName);
+            this.reactor.setTimeout(REACTOR_TIMEOUT);
+            this.reactor.run();
```

Comment:
How long does this run for?

Reply:
It runs until the message is sent, or an exception occurs while opening the connection